### ISSUE TITLE
Problem with "queue:work"

### ISSUE THREAD

**Issue by patryyyck** [spatie/laravel-multitenancy#205]
Hi,

I've a problem with the following command:
`php artisan tenants:artisan --tenant=2 "queue:work"`

I have this error: "**Call to a member function beginTransaction() on null**" in the laravel.log.

I use a multi-database configuration.
Here's my configuration:
.env

DB_CONNECTION=tenant
DB_HOST=mysql
DB_PORT=3306
DB_USERNAME=root
DB_PASSWORD=secret
QUEUE_CONNECTION=database

database.php

    'connections' => [

        'tenant' => [
            'driver' => 'mysql',
            'url' => env('DATABASE_URL'),
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => null,
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'unix_socket' => env('DB_SOCKET', ''),
            'charset' => 'utf8mb4',
            'collation' => 'utf8mb4_unicode_ci',
            'prefix' => '',
            'prefix_indexes' => true,
            'strict' => true,
            'engine' => null,
            'options' => extension_loaded('pdo_mysql') ? array_filter([
                PDO::MYSQL_ATTR_SSL_CA => env('MYSQL_ATTR_SSL_CA'),
            ]) : [],
        ],

        'landlord' => [
            'driver' => 'mysql',
            'url' => env('DATABASE_URL'),
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => 'master_db',
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'unix_socket' => env('DB_SOCKET', ''),
            'charset' => 'utf8mb4',
            'collation' => 'utf8mb4_unicode_ci',
            'prefix' => '',
            'prefix_indexes' => true,
            'strict' => true,
            'engine' => null,
            'options' => extension_loaded('pdo_mysql') ? array_filter([
                PDO::MYSQL_ATTR_SSL_CA => env('MYSQL_ATTR_SSL_CA'),
            ]) : [],
        ],

Everything works except this command.
Thanks for your help.

Regards

**Comment by masterix21** [comment]
Try to replace `DB_CONNECTION=tenant` with `DB_CONNECTION=landlord`

**Comment by patryyyck** [comment]
Ok.
I'll try later.
That means that the queue tables have to be on the landlord connection instead of each tenant?

**Comment by IcarusIT** [comment]
I also have this problem with queues
I also set the connection in landlord but nothing.
always returns null database error

**Comment by sebastiangil** [comment]
I have the same problem

**Comment by masterix21** [comment]
Who has the problem, could create a public repository on GitHub to try, please?

Thanks

**Comment by Ghostscypher** [comment]
So for me, the solution was pretty simple.
1. Set the default connection as that of `landlord`.
2. Make sure that you include this task
 ```php
/*
 * These tasks will be performed to make a tenant current.
 *
 * A valid task is any class that implements Spatie\Multitenancy\Tasks\SwitchTenantTask
 */
'switch_tenant_tasks' => [
    Spatie\Multitenancy\Tasks\SwitchTenantDatabaseTask::class,
],
```
3. For those using the database as the queue driver move the queue migrations to the landlord migrations instead.
4. So by default running queue:work should use the default connection in this case `landlord`, and switching the database tasks should be done by `SwitchTenantDatabaseTask::class`.

**TLDR**

- Queue connection should be handled by landlord therefore queue migrations for database should be in the landlord migrations
- Running `php artisan tenant:artisan "queue:work --tenant=1"` didn't work as the database connection is always set to `null`.
- Since queues are tenant aware by default you would not run into issues of mixing tenant data.
- Also you can even separate the queues by simply running them with different tenant name
 
Here is an example of how the above can be achieved
Suppose we have 3 tenants, appropriately named
1. tenant_1
2. tenant_2
3. tenant_3

We cab create a queue switching task such as the one shown below. For more about creating tasks read the following documentation [creating custom taks](https://spatie.be/docs/laravel-multitenancy/v2/using-tasks-to-prepare-the-environment/creating-your-own-task/)

```php
namespace App\Tasks;

use Spatie\Multitenancy\Models\Tenant;

class SwitchQueueNameTask implements SwitchTenantTask
{
    public function __construct(protected ?string $originalQueue= null)
    {
        $this->originalQueue??= config('queue.default,queue');
    }

    public function makeCurrent(Tenant $tenant): void
    {
         config()->set('queue.default.queue', $tenant->name);
    }

    public function forgetCurrent(): void
    {
       config()->set('queue.default.queue', $this->originalQueue);   
    }

}
```
Then add it to your task list as shown below, (Assuming you've already made the `App\Tasks\` folder autodiscoverable)

 ```php
.....
/*
 * These tasks will be performed to make a tenant current.
 *
 * A valid task is any class that implements Spatie\Multitenancy\Tasks\SwitchTenantTask
 */
'switch_tenant_tasks' => [
    Spatie\Multitenancy\Tasks\SwitchTenantDatabaseTask::class,
    App\Tasks\SwitchQueueNameTask::class,
   ......
],
....
```

Then you can run the queues as follows

`php artisan queue:work ---name=tenant_1`
`php artisan queue:work ---name=tenant_2`
`php artisan queue:work ---name=tenant_3`

Therefore each tenant has their own queue. This is another way to do it, or you can run all queues in the default one. You can also set the queue connection name via the jobs itself. But you get the general idea.

Hope this helps.

**Comment by masterix21** [comment]
Thanks, @Ghostscypher, for your share. I'm closing the issue because it isn't related to a package bug, but feel free to continue to talk about it.

**Comment by skoontastic** [comment]
Just to leave this here. I fought with this for a bit, and realized I forgot to set the models involved to use the tenant connection. So make sure this is in your models:

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Spatie\Multitenancy\Models\Concerns\UsesTenantConnection;

class MyModel extends Model
{
    use HasFactory, UsesTenantConnection;
}
```
